### PR TITLE
Add `ruff`action and configuration

### PR DESCRIPTION
**Pull request by flferretti**
This PR adds the `ruff` action to CIs and pre-commit, together with the configuration that has been added to the `pyproject.toml`. Finally, a full lint using the current version has been performed.

----
📚 Documentation preview 📚: https://jaxsim--114.org.readthedocs.build//114/

### CODE REVIEW

**Comment by diegoferigo** [comment]
Is it ok for you @flferretti to wait for #108 to be ready? Right now, it is easier to rebase this PR on top of that one instead of the opposite.

File: pyproject.toml [hunk 1]
```diff
@@ -14,12 +14,56 @@ version_file = "src/jaxsim/_version.py"
...
+  ".ruff_cache",
+  ".vscode",
+  ".devcontainer",
+  "__pycache__",
 ]
+preview = true
+
+target-version = "py311"
+
+[tool.ruff.lint]
```
Comment (diegoferigo) [inline pyproject.toml:40]:
```suggestion
# https://docs.astral.sh/ruff/rules/
[tool.ruff.lint]
```

File: pyproject.toml [hunk 2]
```diff
@@ -14,12 +14,56 @@ version_file = "src/jaxsim/_version.py"
...
+]
+
+[tool.ruff.lint.per-file-ignores]
+# Ignore `E402` (import violations) in all `__init__.py` files
+"**/{tests,docs,tools}/*" = ["E402"]
+"__init__.py" = ["F401"]
+# TODO: Remove these once deprecated code is removed
+"src/jaxsim/integrators/common.py" = ["E712"]
+"src/jaxsim/physics/model/physics_model.py" = ["B023"]
+"src/jaxsim/simulation/simulator_callbacks.py" = ["F821"]
```
Comment (diegoferigo) [inline pyproject.toml:69]: These two files are no longer in `main`.

File: pyproject.toml [hunk 3]
```diff
@@ -14,12 +14,56 @@ version_file = "src/jaxsim/_version.py"
...
+  "W",
+  "YTT",
+]
+
+[tool.ruff.lint.per-file-ignores]
+# Ignore `E402` (import violations) in all `__init__.py` files
+"**/{tests,docs,tools}/*" = ["E402"]
+"__init__.py" = ["F401"]
+# TODO: Remove these once deprecated code is removed
+"src/jaxsim/integrators/common.py" = ["E712"]
```
Comment (diegoferigo) [inline pyproject.toml:67]: Is this problem easy to fix?

File: src/jaxsim/api/data.py [hunk 4]
```diff
@@ -417,7 +417,7 @@ def base_velocity(self) -> jtp.Vector:
     def generalized_position(self) -> tuple[jtp.Matrix, jtp.Vector]:
         """
         Get the generalized position
-        :math:`\mathbf{q} = ({}^W \mathbf{H}_B, \mathbf{s}) \in \text{SO}(3) \times \mathbb{R}^n`.
+        :math:`\\mathbf{q} = ({}^W \\mathbf{H}_B, \\mathbf{s}) \\in \text{SO}(3) \times \\mathbb{R}^n`.
```
Comment (diegoferigo) [inline src/jaxsim/api/data.py:420]: Can this be fixed using `r"""` as in other docstrings?

File: tests/conftest.py [hunk 5]
```diff
@@ -176,14 +175,12 @@ def jaxsim_model_ergocub() -> js.model.JaxSimModel:
...
 
-    os_environ_original = os.environ.copy()
-
     try:
         os.environ["ROBOT_DESCRIPTION_COMMIT"] = "v0.7.1"
 
         import robot_descriptions.ergocub_description
     finally:
-        os.environ = os_environ_original
+        os.environ.clear()
```
Comment (diegoferigo) [inline tests/conftest.py:183]: Doesn't `clear()` delete the entire dictionary? Here I want to restore the system one.

**Review by diegoferigo — CHANGES_REQUESTED** [review]
Few questions and minor changes.

File: pyproject.toml (hunk 3 above)
Comment (flferretti) [reply pyproject.toml:67]: Potentially yes, the error raised is:

```python
src/jaxsim/integrators/common.py:532:36: E712 Avoid equality comparisons to `True`; use `if rows_of_A_with_fsal:` for truth checks
    |
530 |         # possibly intermediate kᵢ derivative).
531 |         # Note that if multiple rows match (it should not), we return the first match.
532 |         return True, int(jnp.where(rows_of_A_with_fsal == True)[0].tolist()[0])
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ E712
    |
    = help: Replace with `rows_of_A_with_fsal`
```

File: tests/conftest.py (hunk 5 above)
Comment (flferretti) [reply tests/conftest.py:183]: `clear()` deletes the environment variable defined during the process started by Python, we could eventually think to use `subprocess.Popen` to pass a custom env keys set to a subprocess

Comment (diegoferigo) [reply tests/conftest.py:183]: It doesn't seem that it works as you described it:

```
❯ ipython
Python 3.11.8 | packaged by conda-forge | (main, Feb 16 2024, 20:53:32) [GCC 12.3.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.22.2 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import os

In [2]: os.environ["HOME"]
Out[2]: '/home/dferigo'

In [3]: os.environ["FOO"] = "bar"

In [4]: os.environ.clear()

In [5]: os.environ["HOME"]
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
Cell In[5], line 1
----> 1 os.environ["HOME"]

File <frozen os>:679, in __getitem__(self, key)

KeyError: 'HOME'
```

What's the problem of the previous logic?

Comment (diegoferigo) [reply tests/conftest.py:183]: Ow ok now I understand the problem. What about the following logic?

```python
    try:
        os.environ["ROBOT_DESCRIPTION_COMMIT"] = "v0.7.1"

        import robot_descriptions.ergocub_description
    finally:
        _ = os.environ.pop("ROBOT_DESCRIPTION_COMMIT", None)
```

Comment (flferretti) [reply tests/conftest.py:183]: Sure, it makes sense

Comment (flferretti) [reply tests/conftest.py]: Done in https://github.com/ami-iit/jaxsim/pull/114/commits/4cd9358b897b1efdd9621b0c32d31897920a0655

File: src/jaxsim/api/data.py (hunk 4 above)
Comment (flferretti) [reply src/jaxsim/api/data.py:420]: Done in https://github.com/ami-iit/jaxsim/pull/114/commits/f7c63e4ff1a69bb6efd5d40b15544994c938e2ee

File: pyproject.toml (hunk 3 above)
Comment (flferretti) [reply pyproject.toml]: Done in https://github.com/ami-iit/jaxsim/pull/114/commits/f7c63e4ff1a69bb6efd5d40b15544994c938e2ee

File: pyproject.toml (hunk 2 above)
Comment (flferretti) [reply pyproject.toml]: Thanks, done in https://github.com/ami-iit/jaxsim/pull/114/commits/f7c63e4ff1a69bb6efd5d40b15544994c938e2ee

File: pyproject.toml (hunk 1 above)
Comment (flferretti) [reply pyproject.toml:41]: Thanks, done in https://github.com/ami-iit/jaxsim/pull/114/commits/f7c63e4ff1a69bb6efd5d40b15544994c938e2ee

**Review by diegoferigo — APPROVED** [review]

**Pull request merged**